common/darwin/Classes/AudioUtils.m

```diff
@@ -8,73 +8,111 @@ @implementation AudioUtils
+        [session setCategory:config.category
+                 withOptions:config.categoryOptions
+                       error:nil];
+        [session setMode:config.mode error:nil];
+        [session unlockForConfiguration];
+    }
 #endif
 }
 
 + (BOOL)selectAudioInput:(AVAudioSessionPort)type {
 #if TARGET_OS_IPHONE
-  RTCAudioSession *rtcSession = [RTCAudioSession sharedInstance];
-  AVAudioSessionPortDescription *inputPort = nil;
-  for (AVAudioSessionPortDescription *port in rtcSession.session.availableInputs) {
-    if ([port.portType isEqualToString:type]) {
-      inputPort = port;
-      break;
+    RTCAudioSession *rtcSession = [RTCAudioSession sharedInstance];
+    AVAudioSessionPortDescription *inputPort = nil;
+    for (AVAudioSessionPortDescription *port in rtcSession.session.availableInputs) {
+        if ([port.portType isEqualToString:type]) {
+            inputPort = port;
+            break;
+        }
     }
-  }
-  if (inputPort != nil) {
-    NSError *errOut = nil;
-    [rtcSession lockForConfiguration];
-    [rtcSession setPreferredInput:inputPort error:&errOut];
-    [rtcSession unlockForConfiguration];
-    if(errOut != nil) {
-      return NO;
+    if (inputPort != nil) {
+        NSError *errOut = nil;
+        [rtcSession lockForConfiguration];
+        [rtcSession setPreferredInput:inputPort error:&errOut];
+        [rtcSession unlockForConfiguration];
+        if(errOut != nil) {
+            return NO;
+        }
+        return YES;
     }
-    return YES;
-  }
 #endif
-  return NO;
+    return NO;
 }
 
 + (void)setSpeakerphoneOn:(BOOL)enable {
 #if TARGET_OS_IPHONE
     RTCAudioSession *session = [RTCAudioSession sharedInstance];
+    RTCAudioSessionConfiguration *config = [RTCAudioSessionConfiguration webRTCConfiguration];
     [session lockForConfiguration];
-    [session setCategory:AVAudioSessionCategoryPlayAndRecord
-                  withOptions:enable ? AVAudioSessionCategoryOptionDefaultToSpeaker
-                  :
-                  AVAudioSessionCategoryOptionAllowBluetooth | AVAudioSessionCategoryOptionAllowBluetoothA2DP
-                    error:nil];
+    NSError *error = nil;
+    if(!enable) {
+        [session setCategory:config.category
+                 withOptions:config.categoryOptions
+                       error:&error];
+        [session setPreferredOutputNumberOfChannels:0 error:nil];
+        BOOL success = [session setActive:YES error:&error];
+        if (!success) NSLog(@"Audio session override failed: %@", error);
+        else NSLog(@"AudioSession override via Earpiece/Headset is successful ");
+    } else {
+        BOOL success = [session setCategory:config.category
+                                withOptions:AVAudioSessionCategoryOptionDefaultToSpeaker
+                                      error:&error];
+        [session setPreferredOutputNumberOfChannels:0 error:nil];
+        if (!success)  NSLog(@"Port override failed due to: %@", error);
```

Comment:
It is a bit off topic according to MR, but what you think about apply [`Clang-Format`](https://clang.llvm.org/docs/ClangFormat.html) for whole project? It is support all necessary languages - _C/C++/Java/JavaScript/JSON/Objective-C/Protobuf/C#_. And will simplify all this dozen of formatting issues - as double space in this line.

I decided to create separate [discussion](https://github.com/flutter-webrtc/flutter-webrtc/discussions/1125) for this :).